py/runtime.h

```diff
@@ -109,6 +110,46 @@ bool mp_sched_schedule(mp_obj_t function, mp_obj_t arg);
 bool mp_sched_schedule_node(mp_sched_node_t *node, mp_sched_callback_t callback);
 #endif
 
+// Handles any pending MicroPython events without waiting for an interrupt or
+// event.
+STATIC MP_ALWAYSINLINE inline void mp_event_handle_nowait(void) {
+    #if defined(MICROPY_EVENT_POLL_HOOK_FAST) && !MICROPY_PREVIEW_VERSION_2
+    // For ports still using the old macros
+    MICROPY_EVENT_POLL_HOOK_FAST
+    #else
+    mp_handle_pending(true);
```

Comment:
This is not general enough.  Some ports like esp8266 must do work in addition to `mp_handle_pending()`.  So we need a way to hook into this function with a macro.

I guess for now we could use `MICROPY_EVENT_POLL_HOOK_FAST` for this, but eventually we will need a way to hook.

Reply:
> In the specific case of MICROPY_EVENT_POLL_HOOK_FAST and mp_event_handle_no_wait(), esp8266 port doesn't define this hook at the moment so I think it will keep working much the same?

Ah, I'm with you now that the i2c scan specifically failed over to the "slow" hook if the fast hook wasn't defined.

OK, will check all the ports for this behaviour.

Reply:
Updated esp8266 with a new hook `MICROPY_INTERNAL_EVENT_HOOK` and verified i2c.scan() completes. Do you know if any other ports relied on that fallback logic in i2c? It looks like probably not.

Compared to before, there is a small behaviour change - esp8266 port will call `ets_event_poll()` anywhere else that `mp_event_handle_nowait()` is called. I think this is only modselect.c though, and it seems like a good idea to be processing wifi events in that instance.

Reply:
> Ah, I'm with you now that the i2c scan specifically failed over to the "slow" hook if the fast hook wasn't defined.

Indeed.  The existing behaviour is quite subtle, actually too subtle.  I'm pretty sure only esp8266 relied on it.



> Compared to before, there is a small behaviour change - esp8266 port will call `ets_event_poll()` anywhere else that `mp_event_handle_nowait()` is called. I think this is only modselect.c though, and it seems like a good idea to be processing wifi events in that instance.

That sounds like a good improvement.